File: internal/models/monitoring/oncall.go
package monitoring

import "time"

// OncallSchedule 值班排班表
type OncallSchedule struct {
	ID          uint      `gorm:"primarykey" json:"id"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
	Name        string    `gorm:"size:100;not null;uniqueIndex" json:"name"`
	Description string    `gorm:"size:500" json:"description"`
	Timezone    string    `gorm:"size:50;default:'Asia/Shanghai'" json:"timezone"`
	RotationType string   `gorm:"size:20;default:'weekly'" json:"rotation_type"` // daily / weekly / custom
	Enabled     bool      `gorm:"default:true" json:"enabled"`
	CreatedBy   uint      `gorm:"default:0" json:"created_by"`
}

func (OncallSchedule) TableName() string {
	return "oncall_schedules"
}

// OncallShift 值班班次
type OncallShift struct {
	ID         uint      `gorm:"primarykey" json:"id"`
	CreatedAt  time.Time `json:"created_at"`
	ScheduleID uint      `gorm:"not null;index" json:"schedule_id"`
	UserID     uint      `gorm:"not null;index" json:"user_id"`
	UserName   string    `gorm:"size:100" json:"user_name"`
	StartTime  time.Time `gorm:"not null;index" json:"start_time"`
	EndTime    time.Time `gorm:"not null" json:"end_time"`
	ShiftType  string    `gorm:"size:20;default:'primary'" json:"shift_type"` // primary / backup
}

func (OncallShift) TableName() string {
	return "oncall_shifts"
}

// OncallOverride 值班临时替换
type OncallOverride struct {
	ID              uint      `gorm:"primarykey" json:"id"`
	CreatedAt       time.Time `json:"created_at"`
	ScheduleID      uint      `gorm:"not null;index" json:"schedule_id"`
	OriginalUserID  uint      `gorm:"not null" json:"original_user_id"`
	OriginalUserName string   `gorm:"size:100" json:"original_user_name"`
	OverrideUserID  uint      `gorm:"not null" json:"override_user_id"`
	OverrideUserName string   `gorm:"size:100" json:"override_user_name"`
	StartTime       time.Time `gorm:"not null" json:"start_time"`
	EndTime         time.Time `gorm:"not null" json:"end_time"`
	Reason          string    `gorm:"size:500" json:"reason"`
	CreatedBy       uint      `gorm:"default:0" json:"created_by"`
}

func (OncallOverride) TableName() string {
	return "oncall_overrides"
}

// AlertAssignment 告警分配记录
type AlertAssignment struct {
	ID             uint       `gorm:"primarykey" json:"id"`
	CreatedAt      time.Time  `json:"created_at"`
	UpdatedAt      time.Time  `json:"updated_at"`
	AlertHistoryID uint       `gorm:"not null;index" json:"alert_history_id"`
	AssigneeID     uint       `gorm:"not null;index" json:"assignee_id"`
	AssigneeName   string     `gorm:"size:100" json:"assignee_name"`
	ScheduleID     *uint      `json:"schedule_id"`
	Status         string     `gorm:"size:20;default:'pending'" json:"status"` // pending / claimed / resolved / escalated
	ClaimedAt      *time.Time `json:"claimed_at"`
	ResolvedAt     *time.Time `json:"resolved_at"`
	Comment        string     `gorm:"type:text" json:"comment"`
}

func (AlertAssignment) TableName() string {
	return "alert_assignments"
}

File: web/src/services/oncall.ts
import request from '@/services/api'

export interface OncallSchedule {
  id: number
  created_at?: string
  updated_at?: string
  name: string
  description: string
  timezone: string
  rotation_type: string
  enabled: boolean
  created_by: number
}

export interface OncallShift {
  id: number
  schedule_id: number
  user_id: number
  user_name: string
  start_time: string
  end_time: string
  shift_type: string
}

export interface OncallOverride {
  id: number
  schedule_id: number
  original_user_id: number
  original_user_name: string
  override_user_id: number
  override_user_name: string
  start_time: string
  end_time: string
  reason: string
  created_by: number
}

export interface AlertAssignment {
  id: number
  alert_history_id: number
  assignee_id: number
  assignee_name: string
  schedule_id?: number
  status: string
  claimed_at?: string
  resolved_at?: string
  comment?: string
  created_at?: string
}

export const oncallApi = {
  // 排班表
  listSchedules: () => request.get('/oncall/schedules'),
  getSchedule: (id: number) => request.get(`/oncall/schedules/${id}`),
  createSchedule: (data: Partial<OncallSchedule>) => request.post('/oncall/schedules', data),
  updateSchedule: (id: number, data: Partial<OncallSchedule>) => request.put(`/oncall/schedules/${id}`, data),
  deleteSchedule: (id: number) => request.delete(`/oncall/schedules/${id}`),

  // 班次
  listShifts: (scheduleId: number, start?: string, end?: string) =>
    request.get(`/oncall/schedules/${scheduleId}/shifts`, { params: { start, end } }),
  createShift: (scheduleId: number, data: Partial<OncallShift>) =>
    request.post(`/oncall/schedules/${scheduleId}/shifts`, data),
  deleteShift: (shiftId: number) => request.delete(`/oncall/shifts/${shiftId}`),
  generateShifts: (scheduleId: number, data: { user_ids: number[]; user_names: string[]; start_date: string; count: number; type: string }) =>
    request.post(`/oncall/schedules/${scheduleId}/generate`, data),

  // 当前值班
  getCurrentOnCall: (scheduleId: number) => request.get(`/oncall/schedules/${scheduleId}/current`),

  // 临时替换
  listOverrides: (scheduleId: number) => request.get(`/oncall/schedules/${scheduleId}/overrides`),
  createOverride: (scheduleId: number, data: Partial<OncallOverride>) =>
    request.post(`/oncall/schedules/${scheduleId}/overrides`, data),
  deleteOverride: (overrideId: number) => request.delete(`/oncall/overrides/${overrideId}`),

  // 告警分配
  assignAlert: (alertId: number, scheduleId: number) =>
    request.post(`/oncall/alerts/${alertId}/assign`, { schedule_id: scheduleId }),
  claimAlert: (alertId: number) => request.post(`/oncall/alerts/${alertId}/claim`),
  resolveAlert: (alertId: number, comment: string) =>
    request.post(`/oncall/alerts/${alertId}/resolve`, { comment }),
  getAssignment: (alertId: number) => request.get(`/oncall/alerts/${alertId}/assignment`),
  listMyAssignments: (status?: string) => request.get('/oncall/my-assignments', { params: { status } }),
}
